### DISCUSSION TITLE
how to use the register on useEffect in v6?

### DISCUSSION THREAD

**Discussion by ReactGirl** [react-hook-form/react-hook-form#5078] [Q&A]
Hi.

Im having troubles to find a way to register a field by useEffect instead of using the ref. I could do it on v3 butt for v6 the field is not being registered.

I tried the following:

```javascript
useEffect(() => {
    register({ name: "firstName" }, { required: true });
  }, [register]);
```
However it does not work.

I have a code sandbox where I have the full example:
https://codesandbox.io/s/react-hook-form-validation-onblur-onchange-forked-692g9?file=/src/index.js:327-419

Can anyone help me on how I do this register remotely?

**Comment by bluebill1049** [comment]
looks fine to me, and it's working. If you use custom register, then value need to be updated value `setValue`

<img width="1020" alt="Screen Shot 2021-05-08 at 8 47 01 am" src="https://user-images.githubusercontent.com/10513364/117515556-f5b55d80-afd9-11eb-89d5-1a26ed8caaa3.png">

**Reply by ReactGirl** [reply]
@bluebill1049 you are right. It works when you click on the button, but not when you click outside to provoke the onBlue effect.

 In additional to that, I noticed that the "watch" function does not log anything  when the register is done by useEffect. 

<img width="1018" alt="Screenshot 2021-05-07 at 21 01 17" src="https://user-images.githubusercontent.com/29104515/117518814-9161b200-af77-11eb-95c0-fb394e60f375.png">

The behaviour is totally different if I uncomment the line that contains the "ref" inside the component, then everything works.

Any thoughts?

**Reply by bluebill1049** [reply]
it's different, because register at ref we can subscribe to the input, custom register it's like the manual mode, which you have to let the hook form know what's the value of the input.

**Reply by ReactGirl** [reply]
@bluebill1049 

It is a lot better. I tried to add the setValue and the 'watch' started to work, because it knows what is the value of the input. 

It was fixed added the function setValue as you mentioned:
![image](https://user-images.githubusercontent.com/29104515/117540648-82feaf00-afe6-11eb-8857-c38f5ed0c0a4.png)


However, I still missing something that the validation is not trigger when provoke the onBlur effect.


How can I trigger the onBlur validation? 

Any help would be much appreciated.

Sandbox updated:
https://codesandbox.io/s/react-hook-form-validation-onblur-onchange-forked-692g9?file=/src/index.js

**Reply by bluebill1049** [reply]
```
onChange={e => setValue('name', e.taregt.value, {shouldValidate: true })}
onBlur={e => setValue('name', e.taregt.value, {shouldValidate: true })}
```

hope those helps.
